### PR TITLE
Error Overhaul

### PR DESCRIPTION
**Pull request by Justintime50**
# Description

Added the following API errors that all extend `ApiError` which then extends `EasyPostError`:

- ApiError
- EncodingError
- ExternalApiError
- ForbiddenError
- GatewayTimeoutError
- HttpError
- InternalServerError
- InvalidRequestError
- JsonError
- MethodNotAllowedError
- NotFoundError
- PaymentError
- RateLimitError
- RedirectError
- ServiceUnavailableError
- TimeoutError
- UnauthorizedError
- UnknownApiError

Added the following general errors that extend `EasyPostError` which then extends `Exception`:

- EasyPostError
- EndOfPaginationError
- FilteringError
- InvalidObjectError
- InvalidParameterError
- MissingParameterError
- SignatureVerificationError

These errors are either thrown directly or thrown as a part of handling the response from the API.

# Testing

Tests were updated to accomadate new error properties.

# Pull Request Type

Please select the option(s) that are relevant to this PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement (fixing a typo, updating readme, renaming a variable name, etc)

### CODE REVIEW

File: easypost/requestor.py [hunk 1]
```diff
@@ -16,11 +16,47 @@
...
+    TimeoutError,
+    UnauthorizedError,
+    UnknownApiError,
+)
+
+
+STATUS_CODE_TO_ERROR_MAPPING: Dict[int, Any] = {
+    401: UnauthorizedError,
+    402: PaymentError,
+    403: UnauthorizedError,
```
Comment (jchen293) [inline easypost/requestor.py:50]: This should be `ForbiddenError`

File: tests/conftest.py [hunk 2]
```diff
@@ -18,7 +18,6 @@
 EASYPOST_PROD_API_KEY = os.getenv("EASYPOST_PROD_API_KEY")
 PARTNER_USER_PROD_API_KEY = os.getenv("PARTNER_USER_PROD_API_KEY", "123")
 REFERRAL_CUSTOMER_PROD_API_KEY = os.getenv("REFERRAL_CUSTOMER_PROD_API_KEY", "123")
-STRIPE_CONNECT_USER_PROD_API_KEY = os.getenv("STRIPE_CONNECT_USER_PROD_API_KEY", "123")
```
Comment (jchen293) [inline tests/conftest.py:21]: Question: why was this removed?

**Review by jchen293 — CHANGES_REQUESTED** [review]
Overall seems fine, left one blocking change, and also need a test for `TimeoutError`, I believe there are left TODOs for the `easypost_client` test IIRC.

File: tests/conftest.py (hunk 2 above)
Comment (Justintime50) [reply tests/conftest.py:21]: It was added previously when I was fixing the partner tests but wasn't needed after all. The fixture was removed earlier but the constant remained and is unused so I removed it.

File: easypost/requestor.py (hunk 1 above)
Comment (Justintime50) [reply easypost/requestor.py:50]: Nice catch! Fixed

**Review by jchen293 — APPROVED** [review]

**Review by nwithan8 — APPROVED** [review]
